NEP 36 (fair play) (#14779)
numpy/numpy: doc/neps/nep-0036-fair-play.rst

==================
NEP 36 — Fair Play
==================

:Author: Stéfan van der Walt <stefanv@berkeley.edu>
:Status: Draft

> :Type: Standards Track

Reviewer: Should this be "Informational"?

Reviewer: Quoting NEP 0:

```
An Informational NEP describes a NumPy design issue, or provides general guidelines or information to the Python community, but does not propose a new feature. Informational NEPs do not necessarily represent a NumPy community consensus or recommendation, so users and implementers are free to ignore Informational NEPs or follow their advice.

A Process NEP describes a process surrounding NumPy, or proposes a change to (or an event in) a process. Process NEPs are like Standards Track NEPs but apply to areas other than the NumPy language itself. They may propose an implementation, but not to NumPy’s codebase; they require community consensus. Examples include procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in NumPy development. Any meta-NEP is also considered a Process NEP.
```

I think "Process" fits best, but "Informational" is also fine.

:Created: 2019-10-24

> :Resolution: <url> (required for Accepted | Rejected | Withdrawn)

Reviewer: This line should be removed for a non-resolved NEP.

Abstract
--------

This document describes Rules of Play for engaging with the NumPy
project as an external entity.  It discusses usage of the name NumPy,
modified distributions of NumPy, and how to best communicate
with the project when patched versions are deployed.

Companies and developers will know after reading this NEP what kinds

> of behavior the NumPy community consider troublesome, bothersome, and

Reviewer 2: propose "... community would like to see, and which it considers troublesome, ..."

unacceptable.

Motivation
----------

> In December 2018, through the filing of `Issue #12515

Reviewer 2: This wasn't the first time we learned about this. I suggest to turn it around: Third parties redistribute patched versions of NumPy. Sometimes these patches are needed (e.g. build fixes), other times they change or extend NumPy in undesirable ways (e.g. `np.erf` issue)

Reviewer 3: See also: `allow_pickle=False`, where a disputed CVE forced the hand of distro maintainers into shipping numpy with incompatible default arguments ahead of our release

are being made available, and it causes wasted time for our
developers: identifying the release from which the error came, how the
release was modified from the version we provide, and whether we are
able to address the issue (if not, where to relay the report to).

During a community call on XX/XX/XX the community decided to draft a
set of guidelines for companies and external developers to use when
they modify and redistribute NumPy.

Scope
-----

This document aims to define a minimal set of rules that, when
followed, will placate the NumPy developers.  Our hope, however, is
that developers who think that they need to modify NumPy will consider
contributing to the project, or using one of the many mechanisms we
have for allowing our APIs to operate on externally defined objects.

We encourage and exhort developers to first discuss their problem with
the NumPy developers on the `mailing list <>`__ before resorting to
patching and releasing their own version of the library.

Fair Play Rules
---------------

1. Do not reuse the NumPy name for projects not developed by the NumPy
   community.

   At time of writing, there are only two `numpy`-named packages

>    developed by the community: `numpy` and `numpy-financial`.  We ask

Reviewer 2: Not entirely true, e.g. `rnumpy` and `unumpy`. Those are meant for inclusion into the main numpy package, moved under the NumPy GitHub org, or similar. There are many other packages with `numpy` in the name: https://pypi.org/search/?q=numpy. Most of those indeed look undesirable.

     '1.17.2+mycompany.15`

   We understand that minor patches are often required to make a
   library work under a certain distribution.  E.g., Debian may patch
   NumPy so that it searches for optimized BLAS libraries in the
   correct locations.  But we ask that no substantive changes are
   made.

3. Do not extend NumPy's API footprint.

   If you absolutely have to break rule two, please do not add
   additional functions to the namespace.  NumPy's API is already
   quite large, and we are working hard to reduce it where feasible.
   Having additional functions exposed in distributed versions is
   confusing for users and developers alike.

4. *DO* use official mechanism to engage with the API.

   Protocols such as `__array_ufunc__
   <https://numpy.org/neps/nep-0013-ufunc-overrides.html>`__ and
   `__array_function__
   <https://numpy.org/neps/nep-0018-array-function-protocol.html>`__
   were designed to help external packages interact more easily with
   NumPy.  E.g., the latter allows objects from foreign libraries to
   pass through NumPy unharmed.  We actively encourage using any of
   these "officialy sanctioned" mechanisms for overriding or
   interacting with NumPy.

> 

Reviewer 2: Perhaps we should add something like "if those mechanisms aren't sufficient, and you feel like you need to monkeypatch NumPy or modify it in some other way, please discuss it with us on the public mailing list first".

Questions & Answers
-------------------

**Q:** We would like to distribute an optimized version of NumPy that

> utilizes MKL.  You recommend against that, so what are we to do?

Reviewer 2: This question is misleading, we don't recommend against that. Using MKL as the BLAS/LAPACK backend is perfectly fine. That's an example of what we'd like to see.

You're thinking about, e.g. using `mklfft` under the hood without there being a mechanism within NumPy to do so. That requires discussion with us first, and we'd come to some conclusion (e.g. please contribute a backend mechanism for that).

:Type: Informational
:Created: 2019-10-24
:Resolution: Draft

Abstract
--------

This document describes Rules of Play for engaging with the NumPy
project as an external entity.  It discusses usage of the name NumPy,
modified distributions of NumPy, and how to best communicate
with the project when patched versions are deployed.

Companies and developers will know after reading this NEP what kinds
of behavior the community would like to see, and which we consider
troublesome, bothersome, and unacceptable.

Motivation
----------

It has come to our attention that modified versions of NumPy are
distributed by external entities.  Sometimes, the patches are
innocuous, such as when they allow NumPy to function on the target
system.  In other cases, they add new functionality which causes
problems for the community.

For example, in December 2018, through the filing of `Issue #12515
<https://github.com/numpy/numpy/issues/12515>`__, it became known to
the NumPy project that an external company had published a version of
NumPy with an extended footprint.  In this specific instance, a bug
was filed against the `np.erf` function---a function that NumPy itself
had never provided.

> In the months that followed, several conversations was had with the

Reviewer 4: ```suggestion
In the months that followed, several conversations were had with the
```

entity involved, and a better understanding emerged of needs on either

> side.  The entity agreed to publish the patches made against Python,

Reviewer 4: made against python or against numpy? Do they still ship the modified version?

Reviewer 5: `np.erf` and `np.invsqrt` were removed. They still remain in `np.core.umath` in the patched version. 

The plan is to remove extensive patching of NumPy and instead build a scikit-* package which exposes MKL/SVML-backed universal functions.

and a special label was added on the NumPy issue tracker to identify
issues pertaining to this specific distribution.

In another example, a security issue (CVE-2019-6446) was filed against
NumPy.  This forced many distributions to take action, most often by
changing a default keyword value, unfortunately again leading to
multiple versions of the NumPy API being deployed, varying across
distribution.

These are only two of many examples of such modified distributions,
which cause wasted time for our developers: identifying the release
from which the error came, how the release was modified from the
version we provide, and whether we are able to address related issues
(if not, where to relay the report to).

During a community call on XX/XX/XX the community decided to draft a
set of guidelines for companies and external developers to use when
they modify and redistribute NumPy.

Scope
-----

This document aims to define a minimal set of rules that, when

> followed, will placate the NumPy developers.  Our hope, however, is

Reviewer 4: placate is an interesting word choice. Seems a bit negative to me?

Reviewer 5: Perhaps "will minimize unnecessary distractions for the NumPy developers"?

Reviewer 6: Sounds a bit too nice to me ;). I think it is supposed to avoid conflict with developers (and the community), and, in theory, that does include escalating to legal actions. NumPy has a trademark now (although we may have to broadcast that), and it should be clear that enforcing it is a possibility. Although, maybe not in this specific paragraph.

Reviewer 5: In the spirit of Python Zen, "explicit better than implicit", this intended meaning of 'placate' should be spelled out.

Reviewer 2: This is still not addressed, I'm also not a fan of "placate". And yes, it's a bit more than avoiding distractions. The message is "please do not do this". I do not want to allude to anything legal, it's simply common sense/courtesy. How about simply something like:

_This document aims to define a minimal set of rules that we ask everyone who wants to distribute a modified version of NumPy to follow._

Author: The document does not only discuss releases, but I will put in some alternative wording here.

Author: For now, I have:

This document aims to define a minimal set of rules that, when
followed, will be considered good faith efforts in line with the
expectations of the NumPy developers.

Reviewer 2: That sounds better, thanks.

that developers who think that they need to modify NumPy will consider
contributing to the project, or using one of the many mechanisms we
have for allowing our APIs to operate on externally defined objects.

We encourage and exhort developers to first discuss their problem with
the NumPy developers on the `mailing list <>`__ before resorting to
patching and releasing their own version of the library.

Fair Play Rules
---------------

1. Do not reuse the NumPy name for projects not developed by the NumPy
   community.

   At time of writing, there are only a handful of `numpy`-named
   packages developed by the community, including `numpy`,
   `numpy-financial`, and `unumpy`.  We ask that external packages not
   include the phrase `numpy`, i.e., avoid names such as
   `mycompany_numpy`.

>    NumPy is a trademark owned by NumFOCUS.

Reviewer 4: which means using these names probably constitutes a trademark violation in the US, right?

Reviewer 7: > which means using these names probably constitutes a trademark violation in the US, right?

Good point, but I'm not sure adding stronger language would necessarily be an improvement. E.g. something like:

 > NumPy is a trademark owned by NumFOCUS. Unauthorized use of the NumPy name or logo may constitute a violation of this trademark.

but I'd be hesitant to add wording that invites legal discussions in the NEP itself.

Author: I am following up on the NumPy trademark.

==================

> NEP 36 — Fair Play

Reviewer 8: ```suggestion
NEP 36 — Fair play
```
Sentence case for headings.

==================

:Author: Stéfan van der Walt <stefanv@berkeley.edu>
:Status: Draft
:Type: Informational
:Created: 2019-10-24
:Resolution: Draft

Abstract
--------

> This document describes Rules of Play for engaging with the NumPy
> project as an external entity.  It discusses usage of the name NumPy,
> modified distributions of NumPy, and how to best communicate
> with the project when patched versions are deployed.

Reviewer 8: ```suggestion
This document describes Rules of Play for engaging with the NumPy
project as an external entity. 
```
```suggestion
This document sets out Rules of Play for companies and outside developers that engage with the NumPy project. It covers:
- Restrictions on use of the NumPy name
- How and whether to publish a modified distribution
- How to make us aware of patched versions
```
Make it immediately clear who this is addressed to.  Make clear that this is a defining document. Make the scope explicit.

> Companies and developers will know after reading this NEP what kinds
> of behavior the community would like to see, and which we consider
> troublesome, bothersome, and unacceptable.

Reviewer 8: ```suggestion
We identify acceptable approaches from which both parties gain, and unacceptable ones that lead to confusion and wasted effort.
```

Author: I think there's more to it than this.  It is not only that there will be confusion and wasted effort, but that this will antagonize the developers.  And not only because they have to do more work (no-one here seems too scared of that), but because it goes against the philosophy of the project and does not feel like participation in good faith.

Reviewer 8: Those clearly worded sentiments belong in the NEP;  "troublesome, bothersome, and unacceptable" is not the equivalent.

Using my wording a strawman, the NEP can say

> We identify acceptable approaches from which both parties gain, and unacceptable ones that lead to confusion and wasted effort and go against the philosophy of the project.

Insert your preferred wording; antagonism and bad faith can be folded in too. 

At any rate I'd urge against the _-some_ words. Their tone doesn't work in the intended way;  they look like they're waiting for "drat!".

Motivation
----------

> It has come to our attention that modified versions of NumPy are
> distributed by external entities.  Sometimes, the patches are
> innocuous, such as when they allow NumPy to function on the target
> system.  In other cases, they add new functionality which causes
> problems for the community.

Reviewer 8: ```suggestion
We sometimes learn of NumPy versions modified and circulated by outsiders.
These patched versions can cause problems for the NumPy community.
```

> For example, in December 2018, through the filing of `Issue #12515
> <https://github.com/numpy/numpy/issues/12515>`__, it became known to
> the NumPy project that an external company had published a version of
> NumPy with an extended footprint.  In this specific instance, a bug
> was filed against the `np.erf` function---a function that NumPy itself
> had never provided.
> 
> In the months that followed, several conversations were had with the
> entity involved, and a better understanding emerged of needs on either
> side.  The entity agreed to publish the patches made against NumPy,
> and a special label was added on the NumPy issue tracker to identify
> issues pertaining to this specific distribution.

Reviewer 8: - 
```suggestion
- In December 2018, a `bug report <https://github.com/numpy/numpy/issues/12515>`__ was filed against `np.erf` -- a function that didn't exist in the NumPy distribution.  It came to light that a company had published a NumPy version with an extended footprint. After several months of discussion, the company agreed to make its patches public, and we added a label to the NumPy issue tracker to identify issues pertaining to that distribution.
```

> In another example, a security issue (CVE-2019-6446) was filed against
> NumPy.  This forced many distributions to take action, most often by
> changing a default keyword value, unfortunately again leading to
> multiple versions of the NumPy API being deployed, varying across
> distribution.

Reviewer 8: ```suggestion
- In another case, after a security issue (CVE-2019-6446) was filed against
NumPy, distributions put in their own fixes, most often 
 by changing a default keyword value. As a result the NumPy API was
inconsistent across distributions.
```

> These are only two of many examples of such modified distributions,
> which cause wasted time for our developers: identifying the release
> from which the error came, how the release was modified from the
> version we provide, and whether we are able to address related issues
> (if not, where to relay the report to).

Reviewer 8: ```suggestion
In cases like these, our developers waste time identifying the release, locating the alterations, and
determining a course of action: Can we address it? If not, who should?
```

Author: I'll work this in by hand.

> During a community call on `October 16th, 2019 
> <https://github.com/numpy/archive/blob/master/status_meetings/status-2019-10-16.md>`__
> the community decided to draft a
> set of guidelines for companies and external developers to use when
> they modify and redistribute NumPy.

Reviewer 8: ```suggestion
During a community call on `October 16th, 2019 
<https://github.com/numpy/archive/blob/master/status_meetings/status-2019-10-16.md>`__
the community resolved to draft guidelines on distribution of modified NumPy versions.
```

Scope
-----

> This document aims to define a minimal set of rules that, when
> followed, will placate the NumPy developers.  Our hope, however, is
> that developers who think that they need to modify NumPy will consider
> contributing to the project, or using one of the many mechanisms we
> have for allowing our APIs to operate on externally defined objects.

Reviewer 8: ```suggestion
There are preferred alternatives to altering our distribution: contributing to the project, or using one of the many mechanisms we
have for allowing our APIs to operate on externally defined objects. This document sets down minimal rules to be followed when these approaches do not suffice.
```

Author: This changes what I had intended to say, so leaving out for now.

> We encourage and exhort developers to first discuss their problem with
> the NumPy developers on the `mailing list
> <http://numpy-discussion.10968.n7.nabble.com/>`__ before resorting to
> patching and releasing their own version of the library.

Reviewer 8: ```suggestion
In all cases, please talk to us first. Discuss the problem on the `mailing list
<http://numpy-discussion.10968.n7.nabble.com/>`__. We may suggest an
alternative; at minimum, we'll be prepared.
```

> Fair Play Rules

Reviewer 8: ```suggestion
Fair play rules
```
Sentence case.

---------------

1. Do not reuse the NumPy name for projects not developed by the NumPy
   community.

   At time of writing, there are only a handful of `numpy`-named
   packages developed by the community, including `numpy`,
   `numpy-financial`, and `unumpy`.  We ask that external packages not
   include the phrase `numpy`, i.e., avoid names such as
   `mycompany_numpy`.

   NumPy is a trademark owned by NumFOCUS.

2. Do not republish modified versions of NumPy.

   Modified versions of NumPy make it very difficult for the
   developers to address bug reports, since we typically do not know
   which parts of NumPy have been modified.

   If you absolutely have to break this rule (and we implore you not
   to!), then make it absolutely clear in the `__version__` tag that
   you have modified NumPy, e.g.::

     >>> print(np.__version__)
     '1.17.2+mycompany.15`

   We understand that minor patches are often required to make a
   library work under a certain distribution.  E.g., Debian may patch
   NumPy so that it searches for optimized BLAS libraries in the
   correct locations.  But we ask that no substantive changes are
   made.

3. Do not extend NumPy's API footprint.

>    If you absolutely have to break rule two, please do not add
>    additional functions to the namespace.  NumPy's API is already
>    quite large, and we are working hard to reduce it where feasible.
>    Having additional functions exposed in distributed versions is
>    confusing for users and developers alike.

Reviewer 8: ```suggestion
   If you have to break rule 3, do not add
   functions to the namespace.  NumPy's API is already
   large, and we are working to reduce it.
   Exposing additional functions in distributed versions is
   confusing for users and developers alike.
```
Text refers to rule 2, but appears to mean rule 3.

4. *DO* use official mechanism to engage with the API.

   Protocols such as `__array_ufunc__
   <https://numpy.org/neps/nep-0013-ufunc-overrides.html>`__ and
   `__array_function__
   <https://numpy.org/neps/nep-0018-array-function-protocol.html>`__
   were designed to help external packages interact more easily with
   NumPy.  E.g., the latter allows objects from foreign libraries to
   pass through NumPy unharmed.  We actively encourage using any of
   these "officialy sanctioned" mechanisms for overriding or
   interacting with NumPy.

   If these mechanisms are deemed insufficient, please start a
   discussion on the mailing list before monkeypatching NumPy.

> Questions & Answers

Reviewer 8: ```suggestion
Questions and answers
```
Sentence case.

**A:** Please consider including the patches required in the official
NumPy repository.  Not only do we encourage such contributions, but we
already have optimized loops for some platforms available.

**Q:** We would like to ship a much faster version of FFT than NumPy
provides, but NumPy has no mechanism for overriding its FFT routines.
How do we proceed?

**A:** There are two solutions that we approve of: let the users
install your optimizations using a piece of code, such as::

  from my_company_accel import patch_numpy_fft
  patch_numpy_fft()

or have your distribution automatically perform the above, but print a
message to the terminal clearly stating what is happening::

  We are now patching NumPy for optimal performance under MyComp
  Special Platform.  Please direct all bug reports to
  https://mycomp.com/numpy-bugs

If you require additional mechanisms for overriding code, please
discuss this with the development team on the mailing list.

**Q:** We would like to distribute NumPy with faster linear algebra
routines. Are we allowed to do this?

**A:** Yes, this is explicitly supported by linking to a different
version of BLAS.

Discussion
----------

> References and Footnotes

Reviewer 8: ```suggestion
References and footnotes
```
Sentence case.
